File: graphql-spring-boot-starter/src/main/java/com/axxes/graphql/spring/boot/starter/dto/support/DtoMapper.java
package com.axxes.graphql.spring.boot.starter.dto.support;

import com.axxes.graphql.core.model.CompetenceCenter;
import com.axxes.graphql.core.model.Employee;
import com.axxes.graphql.spring.boot.starter.dto.CompetenceCenterDto;
import com.axxes.graphql.spring.boot.starter.dto.EmployeeDto;
import org.springframework.stereotype.Component;

import java.util.Optional;

import static java.util.stream.Collectors.toSet;

@Component
public class DtoMapper {

    public CompetenceCenter toDomain(CompetenceCenterDto competenceCenter) {
        return Optional.ofNullable(competenceCenter)
                .map(cc -> CompetenceCenter.valueOf(cc.name()))
                .orElse(null);
    }

    public CompetenceCenterDto toDto(CompetenceCenter competenceCenter) {
        return CompetenceCenterDto.valueOf(competenceCenter.name());
    }

    public EmployeeDto toDto(Employee employee) {
        return new EmployeeDto(employee.getId(),
                employee.getFirstName(),
                employee.getLastName(),
                employee.getEmail(),
                employee.getCompetenceCenters().stream().map(this::toDto).collect(toSet())
        );
    }
}

File: graphql-kotlin-spring-server/src/main/kotlin/com/axxes/graphql/kotlin/spring/server/dto/support/DtoMapper.kt
package com.axxes.graphql.kotlin.spring.server.dto.support

import com.axxes.graphql.core.model.CompetenceCenter
import com.axxes.graphql.core.model.Employee
import com.axxes.graphql.kotlin.spring.server.dto.CompetenceCenterDto
import com.axxes.graphql.kotlin.spring.server.dto.EmployeeDto
import org.springframework.stereotype.Component

@Component
class DtoMapper {
    fun toDomain(competenceCenter: CompetenceCenterDto?) = competenceCenter?.let { CompetenceCenter.valueOf(it.name) }

    fun toDto(employee: Employee) = with(employee) {
        EmployeeDto(
                id = id,
                firstName = firstName,
                lastName = lastName,
                email = email,
                competenceCenters = competenceCenters.map { it.toDto() }
        )
    }

    private fun CompetenceCenter.toDto() = CompetenceCenterDto.valueOf(name)
}